app/server/lib/FlexServer.ts

```diff
@@ -850,10 +850,17 @@ export class FlexServer implements GristServer {
       baseDomain: this._defaultBaseDomain,
     });
 
-    const isForced = appSettings.section('login').flag('forced').readBool({
+    const forceLogin = appSettings.section('login').flag('forced').readBool({
       envVar: 'GRIST_FORCE_LOGIN',
     });
-    const forcedLoginMiddleware = isForced ? this._redirectToLoginWithoutExceptionsMiddleware : noop;
+    const isAnonPlayground = appSettings.section('login').flag('anonplayground').readBool({
+      envVar: 'GRIST_ANON_PLAYGROUND',
+      defaultValue: true
+    });
+
+    const forcedLoginMiddleware = (forceLogin || !isAnonPlayground) ?
+      this._redirectToLoginWithoutExceptionsMiddleware : noop;
+    const forcedLoginMiddlewareForDocs = forceLogin ? this._redirectToLoginWithoutExceptionsMiddleware : noop;
```

Comment:
That sounds interesting, thanks for this feedback!

That being said, I realize I am not 100% sure of what you call the anonymous playground. I have assumed until now that it's basically allowing the user creating documents though they are not saved until they get authenticated.

> I'd prefer if GRIST_ANON_PLAYGROUND=false specifically disabled the special playground functionality for the anonymous user, while retaining the ability to choose to share sites and workspaces publicly.
>
> This could include changes to makeAnonIntro in app/client/ui/HomeIntro.ts (or a more sweeping denial on the back end of access to the personal site by anon users),

 I guess we'd have to share settings between the backend and the frontend. Could this be a new entry in the returned object of `makeGristConfig` in `app/server/lib/sendAppPage.ts` ?

Regarding the message to place if the option is set to false, I propose something as follow (however, it feels like there miss some icons, it seems stark):
![2023-08-31_18-16](https://github.com/gristlabs/grist-core/assets/371705/1bd23586-16f7-4e6f-b8d9-f3b39119ce15)

> and denial of access by the anon user to endpoints that create "unsaved" documents (POST /api/docs in DocApi.ts comes to mind).

Oh, I realized that docMiddleware was not meant for the DocApi.ts module but for the "document landing page" (in case of separate home and docs workers?). 

I'll attempt to add unit tests. If I am not wrong, I should get inspiration from this piece of code to run tests for environment variables except that I wouldn't call `testDocApi()`: 
https://github.com/gristlabs/grist-core/blob/a18007ae7a3841275507add48e73a8e91a982fbd/test/server/lib/DocApi.ts#L111-L120

I will continue to take a look on Tuesday.

Reply:
Yes, a new entry in the flags sent by sendAppPage.ts as you suggest would make sense. You could also choose to do what `GRIST_FORCE_LOGIN` has been doing, and redirect anon to login/signup for the homepage of the personal site, so you don't have to do any new UI work. Or simply trigger the "access denied" page. Of the three options, the extra UI work feels friendlier.

Yes, `docMiddleware` in `FlexServer` is for the document landing page, not API endpoints. Things like redirecting to login make sense for a web page but not for REST calls.

Hmm I'm not seeing unit tests for /api/docs, I would have expected to see them in DocApi.ts. Perhaps that functionality was only tested end-to-end :( - more testing in this area would be very welcome!

If you place tests within the "testDocApi" function, they will be tested in three scenarios:
 * merged home server / doc worker as single process
 * home server and doc worker as distinct processes
 * a bare doc worker with no home server
The test repetition is important when "plumbing" between home servers and doc workers is being set up or changed.

Reply:
I implemented an API test for this scenario, with the hope it's enough.

> If you place tests within the "testDocApi" function, they will be tested in three scenarios:

I am not sure how I can add this test in `testDocApi()`, as it depends on an environment variable. Thus I called the `setup()` function and added my test for this special case. Otherwise, I propose that I call `testDocApi()` add a new if in this existing test:
https://github.com/gristlabs/grist-core/blob/b9b0632be87830d44b849ce32ae3e65c2808234e/test/server/lib/DocApi.ts#L338-L377

I will continue tomorrow to work on this

Reply:
Right, I adapted the UI. Here is how it looks:

![welcome page as anonymous when GRIST_ANON_PLAYGROUND is set](https://github.com/gristlabs/grist-core/assets/371705/f1d8b3c1-66eb-4501-8470-671b7bd51779)

However, I am still hesitant, as I see the "add new" panel behavior on document is a bit different:

![Add new in doc panel when anonymous](https://github.com/gristlabs/grist-core/assets/371705/5bbddf69-e596-4e48-86db-589c270869f8)

I would understand if you ask me for being consistent with what exist.

Otherwise, I would start writing tests for the UI.